[meta] clippy warnings

Here's what `clippy` currently warns:

```
src/gfx/text.rs:28:40: 28:46 warning: you should put `OpenGL` between ticks in the documentation, #[warn(doc_markdown)] on by default
src/gfx/text.rs:28 /// Handles rendering of debug text to OpenGL
                                                          ^~~~~~
src/gfx/text.rs:28:40: 28:46 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#doc_markdown
src/math/mat.rs:10:5: 10:11 warning: you should put `OpenGL` between ticks in the documentation, #[warn(doc_markdown)] on by default
src/math/mat.rs:10 /// OpenGL.
                       ^~~~~~
src/math/mat.rs:10:5: 10:11 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#doc_markdown
src/math/sphere.rs:177:9: 177:10 warning: 5th binding whose name is just one char, #[warn(many_single_char_names)] on by default
src/math/sphere.rs:177     let a = 0.0;
                               ^
src/math/sphere.rs:177:9: 177:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#many_single_char_names
src/math/sphere.rs:178:9: 178:10 warning: 6th binding whose name is just one char, #[warn(many_single_char_names)] on by default
src/math/sphere.rs:178     let b = 1.0 - a;
                               ^
src/math/sphere.rs:178:9: 178:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#many_single_char_names
src/game/player.rs:115:64: 115:69 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/player.rs:115             if let Some(contact) = level.volume().sweep_sphere(&head, &displacement) {
                                                                                      ^~~~~
src/game/player.rs:115:64: 115:69 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/player.rs:162:60: 162:64 warning: approximate value of `f{32, 64}::PI` found. Consider using it directly, #[warn(approx_constant)] on by default
src/game/player.rs:162         let pitch = clamp(self.camera.pitch() + look[1], (-3.14 / 2.0, 3.14 / 2.0));
                                                                                  ^~~~
src/game/player.rs:162:60: 162:64 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#approx_constant
src/game/player.rs:162:72: 162:76 warning: approximate value of `f{32, 64}::PI` found. Consider using it directly, #[warn(approx_constant)] on by default
src/game/player.rs:162         let pitch = clamp(self.camera.pitch() + look[1], (-3.14 / 2.0, 3.14 / 2.0));
                                                                                              ^~~~
src/game/player.rs:162:72: 162:76 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#approx_constant
src/game/player.rs:205:23: 205:41 warning: this `if` has the same condition as a previous if, #[warn(ifs_same_cond)] on by default
src/game/player.rs:205             } else if contact.time < 1.0 {
                                             ^~~~~~~~~~~~~~~~~~
src/game/player.rs:203:16: 203:34 note: same as this
src/game/player.rs:203             if contact.time < 1.0 {
                                      ^~~~~~~~~~~~~~~~~~
src/game/player.rs:205:23: 205:41 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#ifs_same_cond
src/game/level.rs:51:29: 51:44 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:51                             &wad.metadata(),
                                                 ^~~~~~~~~~~~~~~
src/game/level.rs:51:29: 51:44 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/level.rs:157:56: 157:61 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:157                          .filter(|name| !is_untextured(&name))
                                                                             ^~~~~
src/game/level.rs:157:56: 157:61 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/level.rs:173:5: 188:6 warning: this function has too many arguments (8/7), #[warn(too_many_arguments)] on by default
src/game/level.rs:173     fn build(level: &WadLevel,
                          ^
src/game/level.rs:173:5: 188:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#too_many_arguments
src/game/level.rs:242:9: 249:10 warning: the loop variable `i` is used to index `points`. Consider using `for (i, item) in points.iter().enumerate().skip(2)` or similar iterators, #[warn(needless_range_loop)] on by default
src/game/level.rs:242         for i in 2..points.len() {
                              ^
src/game/level.rs:242:9: 249:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_range_loop
src/game/level.rs:265:9: 272:10 warning: the loop variable `i` is used to index `points`. Consider using `for (i, item) in points.iter().enumerate().skip(2)` or similar iterators, #[warn(needless_range_loop)] on by default
src/game/level.rs:265         for i in 2..points.len() {
                              ^
src/game/level.rs:265:9: 272:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_range_loop
src/game/level.rs:277:9: 280:10 warning: the loop variable `i` is used to index `points`. Consider using `for (i, item) in points.iter().enumerate().skip(1)` or similar iterators, #[warn(needless_range_loop)] on by default
src/game/level.rs:277         for i in 1..points.len() {
                              ^
src/game/level.rs:277:9: 280:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_range_loop
src/game/level.rs:285:9: 288:10 warning: the loop variable `i` is used to index `points`. Consider using `for (i, item) in points.iter().enumerate().skip(1)` or similar iterators, #[warn(needless_range_loop)] on by default
src/game/level.rs:285         for i in 1..points.len() {
                              ^
src/game/level.rs:285:9: 288:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_range_loop
src/game/level.rs:303:9: 306:10 warning: you seem to be trying to use match for destructuring a single pattern. Consider using `if let`, #[warn(single_match)] on by default
src/game/level.rs:303         match marker {
                              ^
src/game/level.rs:303:9: 306:10 help: try this
src/game/level.rs:            if let Marker::StartPos { player: 0 } = marker { *self.start_pos = pos + Vec3f::new(0.0, 0.5, 0.0) }
src/game/level.rs:303:9: 306:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#single_match
src/game/level.rs:324:19: 324:23 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:324             .push(&low, -half_width, 0.0, bounds.size[1], bounds, light_info)
                                        ^~~~
src/game/level.rs:324:19: 324:23 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/level.rs:325:19: 325:23 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:325             .push(&low,
                                        ^~~~
src/game/level.rs:325:19: 325:23 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/level.rs:331:19: 331:24 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:331             .push(&high, -half_width, 0.0, 0.0, bounds, light_info)
                                        ^~~~~
src/game/level.rs:331:19: 331:24 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/level.rs:332:19: 332:23 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:332             .push(&low,
                                        ^~~~
src/game/level.rs:332:19: 332:23 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/level.rs:338:19: 338:24 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:338             .push(&high, half_width, bounds.size[0], 0.0, bounds, light_info)
                                        ^~~~~
src/game/level.rs:338:19: 338:24 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/level.rs:339:19: 339:24 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/game/level.rs:339             .push(&high, -half_width, 0.0, 0.0, bounds, light_info);
                                        ^~~~~
src/game/level.rs:339:19: 339:24 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/game/lights.rs:8:5: 10:6 warning: you should consider deriving a `Default` implementation for `game::lights::LightBuffer`, #[warn(new_without_default_derive)] on by default
src/game/lights.rs:8     pub fn new() -> LightBuffer {
                         ^
src/game/lights.rs:8:5: 10:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive
src/game/lights.rs:8:5: 10:6 help: try this
src/game/lights.rs:       #[derive(Default)]
src/game/game.rs:142:13: 142:82 warning: called `ok().expect()` on a Result value. You can call `expect` directly on the `Result`, #[warn(ok_expect)] on by default
src/game/game.rs:142             frame.finish().ok().expect("Cannot handle context loss currently :(");
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/game/game.rs:142:13: 142:82 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#ok_expect
src/game/world.rs:19:5: 27:6 warning: you should consider deriving a `Default` implementation for `game::world::World`, #[warn(new_without_default_derive)] on by default
src/game/world.rs:19     pub fn new() -> World {
                         ^
src/game/world.rs:19:5: 27:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive
src/game/world.rs:19:5: 27:6 help: try this
src/game/world.rs:       #[derive(Default)]
src/game/world.rs:116:35: 116:47 warning: using `clone` on a `Copy` type, #[warn(clone_on_copy)] on by default
src/game/world.rs:116         self.nodes.push(Node::new(line.clone()));
                                                        ^~~~~~~~~~~~
src/game/world.rs:116:35: 116:47 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#clone_on_copy
src/game/world.rs:122:35: 122:47 warning: using `clone` on a `Copy` type, #[warn(clone_on_copy)] on by default
src/game/world.rs:122         self.nodes.push(Node::new(line.clone()));
                                                        ^~~~~~~~~~~~
src/game/world.rs:122:35: 122:47 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#clone_on_copy
src/gfx/error.rs:34:53: 34:63 warning: this `match` has identical arm bodies, #[warn(match_same_arms)] on by default
src/gfx/error.rs:34             Error::IncompatibleOpenGl(ref inner) => &inner[..],
                                                                        ^~~~~~~~~~
src/gfx/error.rs:33:38: 33:48 note: same as this
src/gfx/error.rs:33             Error::Sdl(ref inner) => &inner[..],
                                                         ^~~~~~~~~~
src/gfx/error.rs:34:53: 34:63 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#match_same_arms
src/gfx/vertex.rs:32:5: 34:6 warning: you should consider deriving a `Default` implementation for `gfx::vertex::FlatBufferBuilder`, #[warn(new_without_default_derive)] on by default
src/gfx/vertex.rs:32     pub fn new() -> Self {
                         ^
src/gfx/vertex.rs:32:5: 34:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive
src/gfx/vertex.rs:32:5: 34:6 help: try this
src/gfx/vertex.rs:       #[derive(Default)]
src/gfx/vertex.rs:58:5: 60:6 warning: you should consider deriving a `Default` implementation for `gfx::vertex::WallBufferBuilder`, #[warn(new_without_default_derive)] on by default
src/gfx/vertex.rs:58     pub fn new() -> Self {
                         ^
src/gfx/vertex.rs:58:5: 60:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive
src/gfx/vertex.rs:58:5: 60:6 help: try this
src/gfx/vertex.rs:       #[derive(Default)]
src/gfx/vertex.rs:62:5: 82:6 warning: this function has too many arguments (8/7), #[warn(too_many_arguments)] on by default
src/gfx/vertex.rs:62     pub fn push(&mut self,
                         ^
src/gfx/vertex.rs:62:5: 82:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#too_many_arguments
src/gfx/vertex.rs:107:5: 109:6 warning: you should consider deriving a `Default` implementation for `gfx::vertex::DecorBufferBuilder`, #[warn(new_without_default_derive)] on by default
src/gfx/vertex.rs:107     pub fn new() -> Self {
                          ^
src/gfx/vertex.rs:107:5: 109:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive
src/gfx/vertex.rs:107:5: 109:6 help: try this
src/gfx/vertex.rs:        #[derive(Default)]
src/gfx/vertex.rs:147:5: 149:6 warning: you should consider deriving a `Default` implementation for `gfx::vertex::SkyBufferBuilder`, #[warn(new_without_default_derive)] on by default
src/gfx/vertex.rs:147     pub fn new() -> Self {
                          ^
src/gfx/vertex.rs:147:5: 149:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive
src/gfx/vertex.rs:147:5: 149:6 help: try this
src/gfx/vertex.rs:        #[derive(Default)]
src/math/mat.rs:25:5: 31:6 warning: this function has too many arguments (16/7), #[warn(too_many_arguments)] on by default
src/math/mat.rs:25     pub fn new(m00: Scalar, m01: Scalar, m02: Scalar, m03: Scalar,
                       ^
src/math/mat.rs:25:5: 31:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#too_many_arguments
src/math/sphere.rs:68:9: 115:10 warning: the loop variable `i` is used to index `triangle`. Consider using `for (i, item) in triangle.iter().enumerate().take(3)` or similar iterators, #[warn(needless_range_loop)] on by default
src/math/sphere.rs:68         for i in 0..3 {
                              ^
src/math/sphere.rs:68:9: 115:10 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_range_loop
src/wad/name.rs:122:9: 122:80 warning: called `ok().expect()` on a Result value. You can call `expect` directly on the `Result`, #[warn(ok_expect)] on by default
src/wad/name.rs:122         str::from_utf8(self.deref()).ok().expect("wad name is not valid utf-8")
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/wad/name.rs:122:9: 122:80 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#ok_expect
src/wad/archive.rs:98:9: 98:45 warning: you seem to be using .map() to clone the contents of an Option, consider using `.cloned()`, #[warn(map_clone)] on by default
src/wad/archive.rs:98         self.index_map.get(name).map(|x| *x)
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/wad/archive.rs:98:9: 98:45 help: try
self.index_map.get(name).cloned()
src/wad/archive.rs:98:9: 98:45 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#map_clone
src/wad/archive.rs:105:9: 106:63 warning: use of `ok_or` followed by a function call, #[warn(or_fun_call)] on by default
src/wad/archive.rs:105         self.named_lump_index(name)
                               ^
src/wad/archive.rs:105:9: 106:63 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#or_fun_call
src/wad/archive.rs:105:9: 106:63 help: try this
src/wad/archive.rs:            self.named_lump_index(name).ok_or_else(|| MissingRequiredLump(format!("{:?}", name)))
src/wad/image.rs:292:9: 292:55 warning: use of `ok_or` followed by a function call, #[warn(or_fun_call)] on by default
src/wad/image.rs:292         self.ok_or(ImageError(Cow::Borrowed(message)))
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/wad/image.rs:292:9: 292:55 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#or_fun_call
src/wad/image.rs:292:9: 292:55 help: try this
src/wad/image.rs:            self.ok_or_else(|| ImageError(Cow::Borrowed(message)))
src/wad/visitor.rs:14:5: 24:6 warning: this function has too many arguments (9/7), #[warn(too_many_arguments)] on by default
src/wad/visitor.rs:14     fn visit_wall_quad(&mut self,
                          ^
src/wad/visitor.rs:14:5: 24:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#too_many_arguments
src/wad/visitor.rs:14:5: 24:6 warning: this function has too many arguments (9/7), #[warn(too_many_arguments)] on by default
src/wad/visitor.rs:14     fn visit_wall_quad(&mut self,
                          ^
src/wad/visitor.rs:14:5: 24:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#too_many_arguments
src/wad/visitor.rs:157:40: 157:45 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/wad/visitor.rs:157         let partition = partition_line(&root);
                                                              ^~~~~
src/wad/visitor.rs:157:40: 157:45 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/wad/visitor.rs:180:40: 180:45 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/wad/visitor.rs:180         let partition = partition_line(&node);
                                                              ^~~~~
src/wad/visitor.rs:180:40: 180:45 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/wad/visitor.rs:373:31: 379:25 warning: this `else { if .. }` block can be collapsed, #[warn(collapsible_if)] on by default
src/wad/visitor.rs:373                        } else {
                                                     ^
src/wad/visitor.rs:373:31: 379:25 help: try
src/wad/visitor.rs:                           } else if is_untextured(&side.lower_texture) {
src/wad/visitor.rs:        Peg::BottomFloat
src/wad/visitor.rs:    } else {
src/wad/visitor.rs:        Peg::Top
src/wad/visitor.rs:    },
src/wad/visitor.rs:373:31: 379:25 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#collapsible_if
src/wad/visitor.rs:383:5: 478:6 warning: this function has too many arguments (8/7), #[warn(too_many_arguments)] on by default
src/wad/visitor.rs:383     fn wall_quad(&mut self,
                           ^
src/wad/visitor.rs:383:5: 478:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#too_many_arguments
src/wad/visitor.rs:429:60: 429:71 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/wad/visitor.rs:429         let light_info = light_info(&mut self.light_cache, &self.level, sector);
                                                                                  ^~~~~~~~~~~
src/wad/visitor.rs:429:60: 429:71 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/wad/visitor.rs:431:16: 431:30 warning: ==-comparison of f32 or f64 detected. Consider changing this to `(v1[0] - v2[0]).abs() < epsilon` for some suitable value of epsilon. std::f32::EPSILON and std::f64::EPSILON are available., #[warn(float_cmp)] on by default
src/wad/visitor.rs:431             if v1[0] == v2[0] {
                                      ^~~~~~~~~~~~~~
src/wad/visitor.rs:431:16: 431:30 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#float_cmp
src/wad/visitor.rs:434:23: 434:37 warning: ==-comparison of f32 or f64 detected. Consider changing this to `(v1[1] - v2[1]).abs() < epsilon` for some suitable value of epsilon. std::f32::EPSILON and std::f64::EPSILON are available., #[warn(float_cmp)] on by default
src/wad/visitor.rs:434             } else if v1[1] == v2[1] {
                                             ^~~~~~~~~~~~~~
src/wad/visitor.rs:434:23: 434:37 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#float_cmp
src/wad/visitor.rs:481:60: 481:71 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/wad/visitor.rs:481         let light_info = light_info(&mut self.light_cache, &self.level, sector);
                                                                                  ^~~~~~~~~~~
src/wad/visitor.rs:481:60: 481:71 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/wad/visitor.rs:645:68: 645:79 warning: this expression borrows a reference that is immediately dereferenced by the compiler, #[warn(needless_borrow)] on by default
src/wad/visitor.rs:645                                  light_info(&mut self.light_cache, &self.level, sector),
                                                                                          ^~~~~~~~~~~
src/wad/visitor.rs:645:68: 645:79 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_borrow
src/wad/light.rs:32:16: 32:39 warning: ==-comparison of f32 or f64 detected. Consider changing this to `(alt_level - base_level).abs() < epsilon` for some suitable value of epsilon. std::f32::EPSILON and std::f64::EPSILON are available., #[warn(float_cmp)] on by default
src/wad/light.rs:32             if alt_level == base_level {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
src/wad/light.rs:32:16: 32:39 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#float_cmp
src/wad/tex.rs:241:25: 241:94 warning: called `ok().expect()` on a Result value. You can call `expect` directly on the `Result`, #[warn(ok_expect)] on by default
src/wad/tex.rs:241         let mut atlas = Image::new(atlas_size[0], atlas_size[1]).ok().expect("atlas too big");
                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/wad/tex.rs:241:25: 241:94 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#ok_expect
src/wad/tex.rs:392:27: 392:70 warning: called `map(f).unwrap_or(a)` on an Option value. This can be done more directly by calling `map_or(a, f)` instead, #[warn(option_map_unwrap_or)] on by default
src/wad/tex.rs:392         let first_frame = maybe_frames.map(|f| &f[0]).unwrap_or(name);
                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/wad/tex.rs:392:27: 392:70 note: replace `map(|f| &f[0]).unwrap_or(name)` with `map_or(name, |f| &f[0])`
src/wad/tex.rs:392:27: 392:70 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#option_map_unwrap_or
src/main.rs:203:24: 206:69 warning: called `map(f).unwrap_or(a)` on an Option value. This can be done more directly by calling `map_or(a, f)` instead, #[warn(option_map_unwrap_or)] on by default
src/main.rs:203         let filename = Path::new(&args[0])
                                       ^
src/main.rs:203:24: 206:69 note: replace `map(|n| n.to_string_lossy()).unwrap_or(Cow::Borrowed("<cannot determine filename>"))` with `map_or(Cow::Borrowed("<cannot determine filename>"), |n| n.to_string_lossy())`
src/main.rs:203:24: 206:69 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#option_map_unwrap_or
src/main.rs:207:9: 209:50 warning: called `ok().expect()` on a Result value. You can call `expect` directly on the `Result`, #[warn(ok_expect)] on by default
src/main.rs:207         writeln!(io::stderr(), "{}: {}", filename, error)
                        ^
src/main.rs:207:9: 209:50 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#ok_expect
```

To get this output, I used:
- rust nightly (`1.11.0-nightly (ec872dc8a 2016-06-07)`),
- `clippy` version `0.75` (installed with `cargo install clippy`).

Then I ran `cargo clippy` on the root directory.

Note: Because of problems with the compilation of the dependencies, I had to `cargo update` the project.

--- a/src/game/level.rs
+++ b/src/game/level.rs
@@ -51,7 +51,7 @@ impl Level {
         let mut start_pos = Vec3f::zero();
         LevelBuilder::build(
             &level,
-            &wad.metadata(),
+            wad.metadata(),
             textures,
             &texture_maps,
             &mut start_pos,
@@ -168,7 +168,7 @@ fn build_decor_atlas(
             let n2 = WadName::from_bytes(&s);
             n1.into_iter().chain(n2)
         })
-        .filter(|name| !is_untextured(&name))
+        .filter(|name| !is_untextured(name))
         .collect::<Vec<_>>();
     let (TransparentImage { pixels, size }, lookup) =
         textures.build_texture_atlas(tex_names.iter());
@@ -184,6 +184,7 @@ struct LevelBuilder<'a, 'b: 'a> {
 }
 
 impl<'a, 'b: 'a> LevelBuilder<'a, 'b> {
+    #[cfg_attr(feature = "cargo-clippy", allow(too_many_arguments))]
     fn build(
         level: &WadLevel,
         meta: &WadMetadata,
@@ -259,8 +260,8 @@ impl<'a, 'b: 'a> LevelVisitor for LevelBuilder<'a, 'b> {
             self.scene
                 .flats_buffer()
                 .push(&v0, height, light_info, bounds)
-                .push(&v1, height, light_info, bounds)
-                .push(&v2, height, light_info, bounds);
+                .push(v1, height, light_info, bounds)
+                .push(v2, height, light_info, bounds);
         }
     }
 
@@ -277,8 +278,8 @@ impl<'a, 'b: 'a> LevelVisitor for LevelBuilder<'a, 'b> {
         for (v1, v2) in vertices.iter().zip(vertices.iter().skip(1)) {
             self.scene
                 .flats_buffer()
-                .push(&v2, height, light_info, bounds)
-                .push(&v1, height, light_info, bounds)
+                .push(v2, height, light_info, bounds)
+                .push(v1, height, light_info, bounds)
                 .push(&v0, height, light_info, bounds);
         }
     }
@@ -289,8 +290,8 @@ impl<'a, 'b: 'a> LevelVisitor for LevelBuilder<'a, 'b> {
             self.scene
                 .sky_buffer()
                 .push(&v0, height)
-                .push(&v1, height)
-                .push(&v2, height);
+                .push(v1, height)
+                .push(v2, height);
         }
     }
 
@@ -299,8 +300,8 @@ impl<'a, 'b: 'a> LevelVisitor for LevelBuilder<'a, 'b> {
         for (v1, v2) in vertices.iter().skip(1).zip(vertices.iter().skip(2)) {
             self.scene
                 .sky_buffer()
-                .push(&v2, height)
-                .push(&v1, height)
+                .push(v2, height)
+                .push(v1, height)
                 .push(&v0, height);
         }
     }
@@ -336,25 +337,25 @@ fn visit_decor(&mut self, &Decor { low, high, half_width, light_info, tex_name }
         };
         self.scene
             .decors_buffer()
-            .push(&low, -half_width, 0.0, bounds.size[1], bounds, light_info)
+            .push(low, -half_width, 0.0, bounds.size[1], bounds, light_info)
             .push(
-                &low,
+                low,
                 half_width,
                 bounds.size[0],
                 bounds.size[1],
                 bounds,
                 light_info,
             )
-            .push(&high, -half_width, 0.0, 0.0, bounds, light_info)
+            .push(high, -half_width, 0.0, 0.0, bounds, light_info)
             .push(
-                &low,
+                low,
                 half_width,
                 bounds.size[0],
                 bounds.size[1],
                 bounds,
                 light_info,
             )
-            .push(&high, half_width, bounds.size[0], 0.0, bounds, light_info)
-            .push(&high, -half_width, 0.0, 0.0, bounds, light_info);
+            .push(high, half_width, bounds.size[0], 0.0, bounds, light_info)
+            .push(high, -half_width, 0.0, 0.0, bounds, light_info);
     }
 }

--- a/src/game/mod.rs
+++ b/src/game/mod.rs
@@ -1,6 +1,7 @@
 pub mod camera;
 pub mod ctrl;
-pub mod game;
+#[cfg_attr(feature = "cargo-clippy", allow(module_inception))]
+mod game;
 pub mod level;
 pub mod lights;
 pub mod player;

--- a/src/game/player.rs
+++ b/src/game/player.rs
@@ -123,7 +123,7 @@ impl Player {
         let mut time_left = delta_time;
         for _ in 0..100 {
             let displacement = self.velocity * time_left;
-            if let Some(contact) = level.volume().sweep_sphere(&head, &displacement) {
+            if let Some(contact) = level.volume().sweep_sphere(head, &displacement) {
                 let adjusted_time = contact.time - 0.001 / displacement.norm();
                 if adjusted_time < 1.0 {
                     let time = clamp(contact.time, (0.0, 1.0));

--- a/src/gfx/scene.rs
+++ b/src/gfx/scene.rs
@@ -231,8 +231,8 @@ impl<'window> SceneBuilder<'window> {
             walls_texture: self.walls_texture.expect(
                 "missing walls texture from SceneBuilder",
             ),
-            flats_buffer: self.flats_buffer.build(&self.window)?,
-            walls_buffer: self.walls_buffer.build(&self.window)?,
+            flats_buffer: self.flats_buffer.build(self.window)?,
+            walls_buffer: self.walls_buffer.build(self.window)?,
 
             sprite_program: self.sprite_program.expect(
                 "missing sprite program from SceneBuilder",
@@ -240,7 +240,7 @@ impl<'window> SceneBuilder<'window> {
             decors_texture: self.decors_texture.expect(
                 "missing decors texture from SceneBuilder",
             ),
-            decors_buffer: self.decors_buffer.build(&self.window)?,
+            decors_buffer: self.decors_buffer.build(self.window)?,
         })
     }
 
@@ -368,7 +368,7 @@ impl<'scene> StaticStep<'scene> {
             .draw(
                 self.buffer,
                 NoIndices(PrimitiveType::TrianglesList),
-                &self.program,
+                self.program,
                 &self,
                 &self.scene.draw_params,
             )

--- a/src/gfx/vertex.rs
+++ b/src/gfx/vertex.rs
@@ -73,6 +73,7 @@ impl WallBufferBuilder {
         WallBufferBuilder(Vec::with_capacity(256))
     }
 
+    #[cfg_attr(feature = "cargo-clippy", allow(too_many_arguments))]
     pub fn push(
         &mut self,
         xz: &Vec2f,

--- a/src/main.rs
+++ b/src/main.rs
@@ -225,7 +225,7 @@ fn main() {
     use std::path::Path;
 
     if let Err(error) = run() {
-        let program = env::args().next().unwrap_or(String::new());
+        let program = env::args().next().unwrap_or_default();
         let filename = Path::new(&program).file_name().map_or_else(
             || {
                 Cow::Borrowed("<cannot determine filename>")

--- a/src/math/mat.rs
+++ b/src/math/mat.rs
@@ -21,6 +21,7 @@ pub struct Mat4 {
 
 impl Mat4 {
     #[cfg_attr(rustfmt, rustfmt_skip)]
+    #[cfg_attr(feature = "cargo-clippy", allow(too_many_arguments))]
     pub fn new(m00: Scalar, m01: Scalar, m02: Scalar, m03: Scalar,
                m10: Scalar, m11: Scalar, m12: Scalar, m13: Scalar,
                m20: Scalar, m21: Scalar, m22: Scalar, m23: Scalar,
@@ -134,11 +135,11 @@ impl fmt::Debug for Mat4 {
     }
 }
 
-impl<'a, 'b> Mul<&'a Mat4> for &'b Mat4 {
+impl Mul<Mat4> for Mat4 {
     type Output = Mat4;
 
     #[cfg_attr(rustfmt, rustfmt_skip)]
-    fn mul(self, rhs: &'a Mat4) -> Mat4 {
+    fn mul(self, rhs: Mat4) -> Mat4 {
         let l = &self.data;
         let r = &rhs.data;
         Mat4 {
@@ -165,11 +166,11 @@ impl<'a, 'b> Mul<&'a Mat4> for &'b Mat4 {
     }
 }
 
-impl<'a, 'b> Add<&'a Mat4> for &'b Mat4 {
+impl Add<Mat4> for Mat4 {
     type Output = Mat4;
 
     #[cfg_attr(rustfmt, rustfmt_skip)]
-    fn add(self, rhs: &'a Mat4) -> Mat4 {
+    fn add(self, rhs: Mat4) -> Mat4 {
         let l = &self.data;
         let r = &rhs.data;
         Mat4 {
@@ -181,11 +182,11 @@ impl<'a, 'b> Add<&'a Mat4> for &'b Mat4 {
     }
 }
 
-impl<'a, 'b> Sub<&'a Mat4> for &'b Mat4 {
+impl Sub<Mat4> for Mat4 {
     type Output = Mat4;
 
     #[cfg_attr(rustfmt, rustfmt_skip)]
-    fn sub(self, rhs: &'a Mat4) -> Mat4 {
+    fn sub(self, rhs: Mat4) -> Mat4 {
         let l = &self.data;
         let r = &rhs.data;
         Mat4 {
@@ -197,29 +198,30 @@ impl<'a, 'b> Sub<&'a Mat4> for &'b Mat4 {
     }
 }
 
-impl Mul<Mat4> for Mat4 {
+impl<'a, 'b> Mul<&'a Mat4> for &'b Mat4 {
     type Output = Mat4;
 
-    fn mul(self, rhs: Mat4) -> Mat4 {
-        &self * &rhs
+    fn mul(self, rhs: &'a Mat4) -> Mat4 {
+        *self * *rhs
     }
 }
 
-impl Add<Mat4> for Mat4 {
+impl<'a, 'b> Add<&'a Mat4> for &'b Mat4 {
     type Output = Mat4;
 
-    fn add(self, rhs: Mat4) -> Mat4 {
-        &self + &rhs
+    fn add(self, rhs: &'a Mat4) -> Mat4 {
+        *self + *rhs
     }
 }
 
-impl Sub<Mat4> for Mat4 {
+impl<'a, 'b> Sub<&'a Mat4> for &'b Mat4 {
     type Output = Mat4;
 
-    fn sub(self, rhs: Mat4) -> Mat4 {
-        &self - &rhs
+    fn sub(self, rhs: &'a Mat4) -> Mat4 {
+        *self - *rhs
     }
 }
+
 
 impl Index<usize> for Mat4 {
     type Output = [f32];

--- a/src/wad/archive.rs
+++ b/src/wad/archive.rs
@@ -147,7 +147,7 @@ impl Archive {
         let mut file = self.file.borrow_mut();
         let info = &self.lumps[index];
         assert!(info.size > 0);
-        assert!(info.size % mem::size_of::<T>() == 0);
+        assert_eq!(info.size % mem::size_of::<T>(), 0);
         let num_elems = info.size / mem::size_of::<T>();
         file.seek(SeekFrom::Start(info.offset)).in_archive(self)?;
         Ok(file.wad_read_many(num_elems).in_archive(self)?)
@@ -156,7 +156,7 @@ impl Archive {
     pub fn read_lump_single<T: WadReadFrom>(&self, index: usize) -> Result<T> {
         let mut file = self.file.borrow_mut();
         let info = &self.lumps[index];
-        assert!(info.size == mem::size_of::<T>());
+        assert_eq!(info.size, mem::size_of::<T>());
         file.seek(SeekFrom::Start(info.offset)).in_archive(self)?;
         Ok(file.wad_read().in_archive(self)?)
     }

--- a/src/wad/light.rs
+++ b/src/wad/light.rs
@@ -112,7 +112,7 @@ pub enum Contrast {
 }
 
 fn id_to_sync(id: u16) -> f32 {
-    ((id as u64 * 1664525 + 1013904223) & 0xffff) as f32 / 15.0
+    ((id as u64 * 1_664_525 + 1_013_904_223) & 0xffff) as f32 / 15.0
 }
 
 fn light_to_f32(level: LightLevel) -> f32 {

--- a/src/wad/name.rs
+++ b/src/wad/name.rs
@@ -28,9 +28,7 @@ impl WadName {
         for (dest, src) in name.iter_mut().zip(value.iter()) {
             if !src.is_ascii() {
                 debug!("Bailed on non-ascii {}", src);
-                return Err(
-                    ErrorKind::BadWadName(value.iter().cloned().collect()).into(),
-                );
+                return Err(ErrorKind::BadWadName(value.to_owned()).into());
             }
 
             let new_byte = match src.to_ascii_uppercase() {
@@ -42,9 +40,7 @@ impl WadName {
                 }
                 b => {
                     debug!("Bailed on ascii {}", b);
-                    return Err(
-                        ErrorKind::BadWadName(value.iter().cloned().collect()).into(),
-                    );
+                    return Err(ErrorKind::BadWadName(value.to_owned()).into());
                 }
             };
             *dest = new_byte;
@@ -56,9 +52,7 @@ impl WadName {
                 value.len(),
                 !nulled
             );
-            Err(
-                ErrorKind::BadWadName(value.iter().cloned().collect()).into(),
-            )
+            Err(ErrorKind::BadWadName(value.to_owned()).into())
         } else {
             Ok(WadName(name))
         }

--- a/src/wad/tex.rs
+++ b/src/wad/tex.rs
@@ -238,7 +238,7 @@ impl TextureDirectory {
         }
         let atlas_size = atlas_size;
 
-        assert!(positions.len() == entries.len());
+        assert_eq!(positions.len(), entries.len());
         // TODO(cristicbz): This should probably split things into multiple atlases or
         // something, but realistically, I'm never going to implement that.
         let mut atlas = Image::new(atlas_size[0], atlas_size[1]).expect("atlas too big");
@@ -289,7 +289,7 @@ impl TextureDirectory {
             image,
             frame_offset,
             num_frames,
-        } in names.into_iter()
+        } in names
         {
             let offset = Vec2::new(column * 64, row * 64);
             if frame_offset == 0 {
@@ -411,7 +411,7 @@ where
         frames_by_first_frame.insert(first_frame, maybe_frames);
     }
     let mut entries = Vec::with_capacity(frames_by_first_frame.len());
-    for (&name, maybe_frames) in frames_by_first_frame.into_iter() {
+    for (&name, maybe_frames) in frames_by_first_frame {
         match maybe_frames {
             Some(frames) => {
                 for (offset, &frame) in frames.iter().enumerate() {

--- a/src/wad/visitor.rs
+++ b/src/wad/visitor.rs
@@ -170,7 +170,7 @@ impl<'a, V: LevelVisitor> LevelWalker<'a, V> {
                 return;
             }
         };
-        let partition = partition_line(&root);
+        let partition = partition_line(root);
         self.visitor.visit_bsp_root(&partition);
         self.children(root, partition);
         self.visitor.visit_bsp_node_end();
@@ -193,7 +193,7 @@ impl<'a, V: LevelVisitor> LevelWalker<'a, V> {
             warn!("Missing entire node with id {}, skipping.", id);
             return;
         };
-        let partition = partition_line(&node);
+        let partition = partition_line(node);
         self.visitor.visit_bsp_node(&partition, branch);
         self.children(node, partition);
         self.visitor.visit_bsp_node_end();
@@ -269,16 +269,13 @@ impl<'a, V: LevelVisitor> LevelWalker<'a, V> {
                 };
 
                 let dist = |l: &Line2f| l.signed_distance(&point);
-
+                let within_bsp = |d: f32| d >= -BSP_TOLERANCE;
+                let within_seg = |d: f32| d <= SEG_TOLERANCE;
                 // The intersection point must lie both within the BSP volume
                 // and the segs volume.
-                if self.bsp_lines.iter().map(|x| dist(x)).all(|d| {
-                    d >= -BSP_TOLERANCE
-                }) &&
-                    self.subsector_seg_lines.iter().map(dist).all(|d| {
-                        d <= SEG_TOLERANCE
-                    })
-                {
+                let inside_bsp_and_segs = self.bsp_lines.iter().map(&dist).all(within_bsp) &&
+                    self.subsector_seg_lines.iter().map(&dist).all(within_seg);
+                if inside_bsp_and_segs {
                     self.subsector_points.push(point);
                 }
             }
@@ -405,6 +402,7 @@ impl<'a, V: LevelVisitor> LevelWalker<'a, V> {
         );
     }
 
+    #[cfg_attr(feature = "cargo-clippy", allow(too_many_arguments))]
     fn wall_quad(
         &mut self,
         sector: &WadSector,
@@ -457,7 +455,7 @@ impl<'a, V: LevelVisitor> LevelWalker<'a, V> {
         };
 
         let light_info_with_contrast;
-        let light_info = light_info(&mut self.light_cache, &self.level, sector);
+        let light_info = light_info(&mut self.light_cache, self.level, sector);
         let light_info = if light_info.effect.is_none() {
             if (v1[0] - v2[0]).abs() < EPSILON {
                 light_info_with_contrast = light::with_contrast(light_info, Contrast::Brighten);
@@ -507,7 +505,7 @@ impl<'a, V: LevelVisitor> LevelWalker<'a, V> {
     }
 
     fn flat_poly(&mut self, sector: &WadSector) {
-        let light_info = light_info(&mut self.light_cache, &self.level, sector);
+        let light_info = light_info(&mut self.light_cache, self.level, sector);
         let (floor_tex, ceil_tex) = (&sector.floor_texture, &sector.ceiling_texture);
         let (floor_sky, ceil_sky) = (is_sky_flat(floor_tex), is_sky_flat(ceil_tex));
         let floor_y = from_wad_height(if floor_sky {
@@ -698,7 +696,7 @@ impl<'a, V: LevelVisitor> LevelWalker<'a, V> {
             low: &low,
             high: &high,
             half_width: half_width,
-            light_info: light_info(&mut self.light_cache, &self.level, sector),
+            light_info: light_info(&mut self.light_cache, self.level, sector),
             tex_name: &name,
         });
     }
@@ -742,7 +740,7 @@ fn min_max_height(level: &Level) -> (WadCoord, WadCoord) {
         .sectors
         .iter()
         .map(|s| (s.floor_height, s.ceiling_height))
-        .fold((32767, -32768), |(min, max), (f, c)| {
+        .fold((32_767, -32_768), |(min, max), (f, c)| {
             (cmp::min(min, f), cmp::max(max, c))
         });
     (min, max + 32)

